cg/meta/workflow/microsalt.py

```diff
@@ -306,3 +356,98 @@ def get_case_id_from_case(self, unique_id: str) -> Tuple[str, None]:
             raise click.Abort
         case_id = case_obj.internal_id
         return case_id, None
+
+    def microsalt_qc(self, case_id: str, run_dir_path: Path, lims_project: str) -> bool:
+        """Check if given microSALT case passes QC check."""
+        samples: List[Sample] = self.get_samples(case_id=case_id)
+        failed_samples: Dict = {}
+        case_qc: Dict = read_json(file_path=Path(run_dir_path, f"{lims_project}.json"))
+
+        for sample in samples:
+            if not sample.sequenced_at:
+                continue
+            self.qc_sample_check(
+                sample=sample, failed_samples=failed_samples, sample_qc=case_qc[sample.internal_id]
+            )
+
+        return self.qc_case_check(
+            case_id=case_id,
+            failed_samples=failed_samples,
+            samples=samples,
+            run_dir_path=run_dir_path,
+        )
+
+    def qc_case_check(
+        self, case_id: str, samples: List[Sample], failed_samples: Dict, run_dir_path: Path
+    ) -> bool:
+        """Perform the final QC check for a microbial case based on failed samples."""
+        qc_pass: bool = True
+
+        for sample in failed_samples:
+            # Check if negative control failed
+            if sample.control == ControlEnum.negative:
+                qc_pass = False
+            # Check if MWR sample failed
+            if sample.application_version.application.tag == MicrosaltAppTags.MWRNXTR003:
+                qc_pass = False
+
+        # Check if more than 10% of MWX samples failed
+        if len(failed_samples) / len(samples) > MicrosaltQC.QC_PERCENT_THRESHOLD_MWX:
+            qc_pass = False
+
+        if not qc_pass:
+            LOG.warning(
+                f"Case {case_id} failed QC, see {run_dir_path}/QC_done.txt for more information."
+            )
+        else:
+            LOG.info(f"Case {case_id} passed QC.")
+
+        self.create_qc_done_file(
+            run_dir_path=run_dir_path,
+            failed_samples=failed_samples,
+        )
+        return qc_pass
+
+    def create_qc_done_file(self, run_dir_path: Path, failed_samples: Dict) -> None:
+        """Creates a QC_done when a QC check is performed."""
+        with open(run_dir_path.joinpath("QC_done.txt"), "w") as file:
+            for sample_dict in failed_samples.items():
+                file.write(f"{sample_dict[0].internal_id}: {json.dumps(sample_dict[1])} \n")
```

Comment:
use the functions in the io module for writing json files

Reply:
I used it previously and it worked but when I now add what samples fail and on what QC check I had to do it like this. And this file is not a json so I cant use json.dump so instead I use file.write. 

Do you think I should rewrite it so it creates a json file instead and then use the IO function=

Reply:
What happened when you used the IO functions now?

Reply:
Don't remember exactly, but I believe it got too complex when I started posting the failed samples dict in the QC_done file. Since the failed samples dict is nested with other dicts and sample objects I believe the json.dump could not dump sample objects. I tried using the IO function but the could looked more complex than it does now

Reply:
The IO function should be fine with nested dictionaries. I think you can use the internal_id as the dict key and just fetch the sample object from status_db where it's needed.